robot_smach_states/src/robot_smach_states/utility.py

```diff
@@ -351,6 +353,109 @@ def execute(self, userdata=None):
+            "None of the available arms meets all this state machine's requirements: {}".format(arm_requirements)
+    except AssertionError as e:
+        rospy.logerr("Getting arm requirements failed, arm requirements: {}".format(arm_requirements))
+        raise
+
+
+def collect_arm_requirements(state_machine):
+    """ Collects all requirements on the arm of this specific state machine
+
+    :param state_machine: State (machine) for which the requirements need to be collected
+    :return: All arm requirements of the state machine
+    """
+    def update_requirements(state):
+        """ Checks the input state for arm requirements and updates the current arm requirements if necessary
+
+        :param state: Smach state of which arm requirements should be checked against the current arm requirements
+        :return: current arm requirements, updated (if necessary) given the input state
+        """
+        for k, v in state.iteritems():
+            if k not in arm_requirements:
+                arm_requirements[k] = v
+            else:
+                for value in v:
+                    if value not in arm_requirements[k]:
+                        arm_requirements[k] += v
+
+    # Check arm requirements
+    arm_requirements = {}
+
+    if isinstance(state_machine, smach.State):
+        if hasattr(state_machine, "REQUIRED_ARM_PROPERTIES"):
+            update_requirements(state_machine.REQUIRED_ARM_PROPERTIES)
+
+    if isinstance(state_machine, smach.StateMachine):
+        for child_state in state_machine.get_children().itervalues():
+
+            # Check if the child_state is a state_machine (must be done before checking for arm properties!)
+            if isinstance(child_state, smach.StateMachine):
+                child_sm_arm_requirements = collect_arm_requirements(child_state)
+                update_requirements(child_sm_arm_requirements)
+
+            if hasattr(child_state, "REQUIRED_ARM_PROPERTIES"):
+                update_requirements(child_state.REQUIRED_ARM_PROPERTIES)
+
+    rospy.logdebug("These are the collected arm requirements:{}".format(arm_requirements))
+    return arm_requirements
+
+
+def const_resolve(arm_designator, additional_properties):
+    """
+    Resolves the designator after adding properties. Note that the state is not altered.
+    :param arm_designator: ArmDesignator to which properties should be added.
+    :param additional_properties: Dict with the additional properties that are to be added.
+    :return: an arm with the desired properties and state
+    """
+
+    arm_designator_add_props = copy.copy(arm_designator)
+    for k, v in additional_properties.items():
+        if k in arm_designator_add_props.arm_properties:
+            for val in additional_properties[k]:
+                if val not in additional_properties[k]:
+                    arm_designator_add_props.arm_properties[k] += val
+        else:
+            arm_designator_add_props.arm_properties[k] = v
```

Comment:
This is the loop that is very similar and which might be best served as its own function.